File: scripts/custom_packets_demo.cpp
/*
 * Custom Packets Demo Script
 * 
 * This script demonstrates the custom packets system by responding to
 * ping requests from the client with pong responses.
 * 
 * Part of the custom-packets Thorium mod.
 */

#include "ScriptMgr.h"
#include "Player.h"
#include "CustomPacketRead.h"
#include "CustomPacketWrite.h"
#include "CustomPacketHandler.h"
#include "Log.h"

// Custom opcodes - must match client-side (CustomPackets addon)
enum CustomPacketOpcodes
{
    // Client -> Server (CYCM = Custom Client Message)
    CYCM_PING           = 1001,
    CYCM_ECHO           = 1002,
    
    // Server -> Client (CYSM = Custom Server Message)  
    CYSM_PONG           = 2001,
    CYSM_ECHO_RESPONSE  = 2002,
};

class CustomPacketsDemoHandler : public ServerScript
{
public:
    CustomPacketsDemoHandler() : ServerScript("CustomPacketsDemoHandler") { }

    // Called when a custom packet is received from the client
    void OnCustomPacketReceive(Player* player, uint16 opcode, CustomPacketRead* packet) override
    {
        switch (opcode)
        {
            case CYCM_PING:
                HandlePing(player, packet);
                break;
            case CYCM_ECHO:
                HandleEcho(player, packet);
                break;
        }
    }

private:
    void HandlePing(Player* player, CustomPacketRead* packet)
    {
        // Read the timestamp sent by client
        uint32 clientTimestamp = packet->Read<uint32>(0);

        TC_LOG_INFO("custom.packets", "Player {} sent PING with timestamp {}", 
            player->GetName(), clientTimestamp);

        // Build pong response
        CustomPacketWrite response = CreateCustomPacket(CYSM_PONG);
        response.Write<uint32>(clientTimestamp);           // Echo back for latency calculation
        response.Write<uint32>(time(nullptr));             // Server timestamp
        response.WriteString("Pong from TrinityCore!");    // Message
        
        SendCustomPacket(player, response);
    }

    void HandleEcho(Player* player, CustomPacketRead* packet)
    {
        // Read the message from client
        std::string message = packet->ReadString("");

        TC_LOG_INFO("custom.packets", "Player {} sent ECHO: {}", 
            player->GetName(), message);

        // Echo it back with server prefix
        CustomPacketWrite response = CreateCustomPacket(CYSM_ECHO_RESPONSE);
        response.WriteString("Server received: " + message);
        
        SendCustomPacket(player, response);
    }
};

void AddSC_custom_packets_demo()
{
    TC_LOG_INFO("server.loading", "Loading CustomPacketsDemoHandler...");
    new CustomPacketsDemoHandler();
}

File: addons/Interface/AddOns/CustomPacketsDemo/CustomPacketsDemo.lua
-- CustomPackets Demo Addon
-- Demonstrates sending and receiving custom packets

-- Debug: Check if ClientExtensions functions are available
print("DEBUG: CreateCustomPacket type:", type(CreateCustomPacket))
print("DEBUG: OnCustomPacket type:", type(OnCustomPacket))
print("DEBUG: _CLIENT_NETWORK type:", type(_CLIENT_NETWORK))

-- Opcodes must match server-side script
local CYCM_PING = 1001          -- Client -> Server
local CYCM_ECHO = 1002          -- Client -> Server
local CYSM_PONG = 2001          -- Server -> Client
local CYSM_ECHO_RESPONSE = 2002 -- Server -> Client

-- Track pending ping for latency calculation
local pingStartTime = nil

-- ============================================================================
-- Packet Handlers (Server -> Client)
-- ============================================================================

-- Handle pong response from server
OnCustomPacket(CYSM_PONG, function(reader)
    local echoedTimestamp = reader:ReadUInt32(0)
    local serverTimestamp = reader:ReadUInt32(0)
    local message = reader:ReadString("")
    
    -- Calculate round-trip latency
    local latency = 0
    if pingStartTime then
        latency = math.floor((GetTime() - pingStartTime) * 1000)
        pingStartTime = nil
    end
    
    print(string.format("|cFF00FF00[CustomPackets]|r PONG received! Latency: %dms", latency))
    print(string.format("  Server says: %s", message))
end)

-- Handle echo response from server
OnCustomPacket(CYSM_ECHO_RESPONSE, function(reader)
    local message = reader:ReadString("")
    print(string.format("|cFF00FF00[CustomPackets]|r Echo response: %s", message))
end)

-- ============================================================================
-- Slash Commands
-- ============================================================================

SLASH_CPTEST1 = "/cptest"
SLASH_CPTEST2 = "/custompacket"
SlashCmdList["CPTEST"] = function(msg)
    local cmd, arg = msg:match("^(%S+)%s*(.*)$")
    cmd = cmd or msg
    
    if cmd == "ping" then
        -- Send a ping to measure latency
        print("DEBUG: About to create packet")
        pingStartTime = GetTime()
        local packet = CreateCustomPacket(CYCM_PING, 4)
        print("DEBUG: Packet created, about to write")
        packet:WriteUInt32(math.floor(GetTime() * 1000))
        print("DEBUG: Data written, about to send")
        packet:Send()
        print("|cFFFFFF00[CustomPackets]|r Sending PING... (packet sent)")
        
    elseif cmd == "echo" then
        -- Send an echo message
        local message = arg ~= "" and arg or "Hello from WoW client!"
        local packet = CreateCustomPacket(CYCM_ECHO, 0)
        packet:WriteString(message)
        packet:Send()
        print(string.format("|cFFFFFF00[CustomPackets]|r Sending ECHO: %s", message))
        
    else
        print("|cFF00FFFF[CustomPackets Demo]|r Commands:")
        print("  /cptest ping - Send ping, measure latency")
        print("  /cptest echo [message] - Echo a message through server")
    end
end

-- ============================================================================
-- Initialization
-- ============================================================================

print("|cFF00FF00[CustomPackets Demo]|r Loaded. Type /cptest for commands.")
